### PR TITLE
feat: add route /blocks/{blockId}/extrinsics/{extrinsicIndex} 

### PR DESCRIPTION
**Pull request by TarikGul**
## Feat

Added the route `/blocks/{blockId}/extrinsics/{extrinsicsIndex}`

## Changelog 

`src/controllers/blocks/BlocksExtrinsicsController.ts`
Added: `BlocksExtrinsicsController`. This controller uses the `BlocksService` in order to fetch the block, then retrieve the extrinsic. 

`src/services/blocks/BlocksService.ts`
Added: `fetchExtrinsicsByIndex`. Fetch's the extrinsic that is requested in the params. 

`src/services/blocks/BlocksService.spec.ts`
Added a test 'describe'  suite a called `fetchExrinsicsByIndex`.

Additionals:
Added mock data in: `src/services/test-helpers/responses/blocks/blocks789629Extrinsic.json`

## Notes

Inside of the test suite, I had to reuse one line of code between 2 'it' statements. This is because jest was giving me issues with using async in a describe block. It treats it as a nested asynchronous function which will throw a unhandled promise, and abort the test or give inaccurate results. 

closes [#382 ](https://github.com/paritytech/substrate-api-sidecar/issues/382)

### CODE REVIEW

File: docs/src/openapi-v1.yaml [hunk 1]
```diff
@@ -273,6 +273,56 @@ paths:
...
+      - name: blockId
+        in: path
+        description: Block identifier, as the block height or block hash.
+        required: true
+        schema:
+          pattern: a-km-zA-HJ-NP-Z1-9{8,64}
+          type: string
+      - name: extrinsicsIndex
+        in: path
+        description: The index in which the requested extrinsic exists.
```
Comment (emostov) [inline docs/src/openapi-v1.yaml:293]:
```suggestion
        description: The extrinsic's index within the block's body.
```

File: docs/src/openapi-v1.yaml [hunk 2]
```diff
@@ -273,6 +273,56 @@ paths:
             application/json:
               schema:
                 $ref: '#/components/schemas/Error'
+  /blocks/{blockId}/extrinsics/{extrinsicsIndex}:
+    get:
+      tags:
+      - blocks
+      summary: Get an extrinsic by its extrinsicIndex and block height or hash.
```
Comment (emostov) [inline docs/src/openapi-v1.yaml:280]:
```suggestion
      summary: Get an extrinsic by its `Timepoint` (blockId, extrinsicIndex).
```
@danforbes @joepetrowski thoughts?

File: docs/src/openapi-v1.yaml [hunk 3]
```diff
@@ -273,6 +273,56 @@ paths:
...
+      description: Returns a single extrinsic.
+      operationId: getExtrinsicsByExtrsinsicIndex
```
Comment (emostov) [inline docs/src/openapi-v1.yaml:282]:
```suggestion
      operationId: getExtrinsicByTimepoint
```
If we decide to use the `Timepoint` language

File: docs/src/openapi-v1.yaml [hunk 4]
```diff
@@ -273,6 +273,56 @@ paths:
...
+      parameters:
+      - name: blockId
+        in: path
+        description: Block identifier, as the block height or block hash.
+        required: true
+        schema:
+          pattern: a-km-zA-HJ-NP-Z1-9{8,64}
+          type: string
+      - name: extrinsicsIndex
```
Comment (emostov) [inline docs/src/openapi-v1.yaml:291]:
```suggestion
      - name: extrinsicIndex
```

File: docs/src/openapi-v1.yaml [hunk 5]
```diff
@@ -1033,6 +1083,15 @@ components:
...
             extrinsic is signed when reconciling old blocks.
+    ExtrinsicIndex:
+      type: object
+      properties:
+        at:
+          $ref: '#/components/schemas/BlockIdentifiers'
+        extrinsic:
+          $ref: '#/components/schemas/Extrinsic'
+      description: Object returned after keying into the correct extrinsic, 
+        and adding its blockHash, and block height in a `at` field.
```
Comment (emostov) [inline docs/src/openapi-v1.yaml:1094]:
```suggestion
      description: A single extrinsic at a given block.
```

File: src/controllers/blocks/BlocksExtrinsicsController.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+export default class BlocksExtrinsicsController extends AbstractController<BlocksService> {
+	constructor(api: ApiPromise, private readonly options: ControllerOptions) {
+		super(api, '/blocks/:blockId/extrinsics', new BlocksService(api));
+		this.initRoutes();
+	}
+
+	protected initRoutes(): void {
+		this.safeMountAsyncGetHandlers([
+			['/:extrinsicsIndex', this.getExtrinsicsByExtrsinsicIndex],
```
Comment (emostov) [inline src/controllers/blocks/BlocksExtrinsicsController.ts:20]: Should we change all mentions of `extrinsicsIndex` to `extriniscIndex` since we only get a single extrinsic?

File: src/controllers/blocks/BlocksExtrinsicsController.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+		]);
+	}
+
+	/**
+	 *
+	 * @param _req Express Request
+	 * @param res Express Response
+	 */
+	private getExtrinsicsByExtrsinsicIndex: RequestHandler<INumberParam> = async (
```
Comment (emostov) [inline src/controllers/blocks/BlocksExtrinsicsController.ts:29]:
```suggestion
	private getExtrinsicByTimepoint: RequestHandler<INumberParam> = async (
```
?

File: src/controllers/blocks/BlocksExtrinsicsController.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+			checkFinalized: true,
+			queryFinalizedHead,
+			omitFinalizedTag,
+		};
+
+		const block = await this.service.fetchBlock(hash, options);
+
+		this.parseNumberOrThrow(
+			extrinsicsIndex,
+			'ExstrinsicIndex is not a number'
```
Comment (emostov) [inline src/controllers/blocks/BlocksExtrinsicsController.ts:56]:
```suggestion
			'`exstrinsicIndex` uri param is not a number'
```

File: src/services/blocks/BlocksService.ts [hunk 9]
```diff
@@ -249,6 +250,30 @@ export class BlocksService extends AbstractService {
 		};
 	}
 
+	/**
+	 *
+	 * @param block Takes in a block which is the result of fetchBlock
```
Comment (emostov) [inline src/services/blocks/BlocksService.ts:255]:
```suggestion
	 * @param block Takes in a block which is the result of `BlocksService.fetchBlock`
```

File: src/services/blocks/BlocksService.ts [hunk 10]
```diff
@@ -249,6 +250,30 @@ export class BlocksService extends AbstractService {
...
+	 * @param extrinsicsIndex Parameter passed into the request
+	 */
+	fetchExtrinsicsByIndex(
+		block: IBlock,
+		extrinsicsIndex: number
+	): IExtrinsicIndex {
+		if (extrinsicsIndex > block.extrinsics.length - 1) {
+			throw new BadRequest('Requested ExtrinsicIndex does not exist');
```
Comment (emostov) [inline src/services/blocks/BlocksService.ts:263]:
```suggestion
			throw new BadRequest('Requested `extrinsicIndex` does not exist');
```

File: src/controllers/blocks/BlocksExtrinsicsController.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+		);
```
Comment (emostov) [inline src/controllers/blocks/BlocksExtrinsicsController.ts:57]: We could add a test for `extrinsicIndex < 0`

File: src/types/responses/Extrinsic.ts [hunk 12]
```diff
@@ -29,3 +30,11 @@ export interface ISignature {
 	signature: EcdsaSignature | Ed25519Signature | Sr25519Signature;
 	signer: Address;
 }
+
+export interface IExtrinsicIndex {
+	at: {
```
Comment (emostov) [inline src/types/responses/Extrinsic.ts:35]: Should be able to use `IAt` interface here. Because of some legacy code height is a string, so will need to unwrap if its a compact and then `.toString(10)` to ensure it serializes as a base 10.

File: src/controllers/blocks/BlocksExtrinsicsController.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+		const queryFinalizedHead = !this.options.finalizes ? false : true;
+		const omitFinalizedTag = !this.options.finalizes ? true : false;
+
+		const options = {
+			eventDocs: eventDocsArg,
+			extrinsicDocs: extrinsicDocsArg,
+			checkFinalized: true,
+			queryFinalizedHead,
+			omitFinalizedTag,
```
Comment (emostov) [inline src/controllers/blocks/BlocksExtrinsicsController.ts:49]: I think we should hardcode `omitFinalizedTag` too `true` in order to avoid any of the `isFinalizedBlock` logic. Then both `checkFinalized` and `queryFinalizedHead` can be whatever since (I think) they are not used.

**Review by emostov — COMMENTED** [review]
LGTM modulo some nits mostly on just on naming/ docs.

File: docs/src/openapi-v1.yaml (hunk 2 above)
Comment (dvdplm) [reply docs/src/openapi-v1.yaml:280]: Is "Timepoint" and established term in our documentation? If not I think it sounds a bit funny. I found the [`pallet_multisig::Timepoint`](https://substrate.dev/rustdocs/v2.0.0/pallet_multisig/struct.Timepoint.html) but not sure how widely used it is.

File: docs/src/openapi-v1.yaml [hunk 14]
```diff
@@ -1604,4 +1663,4 @@ components:
         application/json:
           schema:
             $ref: '#/components/schemas/Transaction'
-      required: true
+      required: true
```
Comment (dvdplm) [inline docs/src/openapi-v1.yaml:1666]:
```suggestion
      required: true
      
```

File: src/controllers/blocks/BlocksExtrinsicsController.ts (hunk 6 above)
Comment (dvdplm) [reply src/controllers/blocks/BlocksExtrinsicsController.ts:20]: I think so yes; one could argue that it is one index (singular) within a collection of extrinsics and so `extrinsicsIndex` isn't wrong. To my ears `extrinsicIndex` sounds better though.

File: src/controllers/blocks/BlocksExtrinsicsController.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+		const block = await this.service.fetchBlock(hash, options);
+
+		this.parseNumberOrThrow(
+			extrinsicsIndex,
+			'ExstrinsicIndex is not a number'
+		);
+
+		// Change extrinsicIndex from a type string to a number before passing it
+		// into any service.
+		const index = parseInt(extrinsicsIndex, 10);
```
Comment (dvdplm) [inline src/controllers/blocks/BlocksExtrinsicsController.ts:61]: We should probably check for `NaN` here and bail if so?

File: src/services/blocks/BlocksService.spec.ts [hunk 16]
```diff
@@ -358,4 +360,41 @@ describe('BlocksService', () => {
...
+			omitFinalizedTag: false,
+		};
+
+		it('Returns the correct extrinisics object for block 789629', async () => {
+			const block = await blocksService.fetchBlock(
+				blockHash789629,
+				options
+			);
+
+			const extrinsic = blocksService['fetchExtrinsicsByIndex'](block, 0);
```
Comment (dvdplm) [inline src/services/blocks/BlocksService.spec.ts:380]: Would it be better to fetch a non-zero index to prove that we can fetch xts out of order?

File: src/services/blocks/BlocksService.ts [hunk 17]
```diff
@@ -249,6 +250,30 @@ export class BlocksService extends AbstractService {
 		};
 	}
 
+	/**
+	 *
+	 * @param block Takes in a block which is the result of fetchBlock
+	 * @param extrinsicsIndex Parameter passed into the request
+	 */
+	fetchExtrinsicsByIndex(
```
Comment (dvdplm) [inline src/services/blocks/BlocksService.ts:258]: This should be `fetchExtrinsicByIndex` (singular).

File: src/services/test-helpers/responses/blocks/block789629Extrinsic.json [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+                        "class": "Mandatory",
+                        "paysFee": "Yes"
+                    }
+                ]
+            }
+        ],
+        "success": true,
+        "paysFee": false
+    }
+}
```
Comment (dvdplm) [inline src/services/test-helpers/responses/blocks/block789629Extrinsic.json:37]: Most editors can be configured to ensure there is a newline ending each file.

**Review by dvdplm — CHANGES_REQUESTED** [review]
Makes sense to me. I think we should switch to use singular form (`extrinsicIndex` etc), otherwise lgtm.

**Comment by joepetrowski** [comment]
> I think we should switch to use singular form (extrinsicIndex etc)

I agree

File: docs/src/openapi-v1.yaml (hunk 2 above)
Comment (TarikGul) [reply docs/src/openapi-v1.yaml:280]: Yea `Timepoint` does sound a little funny, but that being said a lot of terminology can be weird sometimes, so if this is the correct verbiage I agree with it.

File: src/controllers/blocks/BlocksExtrinsicsController.ts (hunk 6 above)
Comment (TarikGul) [reply src/controllers/blocks/BlocksExtrinsicsController.ts:20]: Yea I agree with you both. I actually had to say it out loud a few times but `extrinsicIndex` definitely makes it sound way clearer.

File: src/controllers/blocks/BlocksExtrinsicsController.ts (hunk 15 above)
Comment (TarikGul) [reply src/controllers/blocks/BlocksExtrinsicsController.ts:61]: Yea I totally agree there is actually a function inside the `AbstractController`  called `parseNumberOrThrow` which in essence takes care of this issue. Thanks for reminding me. :)

File: src/services/blocks/BlocksService.spec.ts (hunk 16 above)
Comment (TarikGul) [reply src/services/blocks/BlocksService.spec.ts:380]: Yea do you think adding another test case would suite useful here, or just changing to a non-zero index?

File: src/services/test-helpers/responses/blocks/block789629Extrinsic.json (hunk 18 above)
Comment (TarikGul) [reply src/services/test-helpers/responses/blocks/block789629Extrinsic.json:37]: Interesting, i thought the linter would catch that. I'll make sure to configure my editor as well.

File: src/types/responses/Extrinsic.ts (hunk 12 above)
Comment (TarikGul) [reply src/types/responses/Extrinsic.ts:35]: Thanks for pointing that out. The beauty of `unwrap`

File: docs/src/openapi-v1.yaml (hunk 2 above)
Comment (joepetrowski) [reply docs/src/openapi-v1.yaml:280]: Yeah, only usage is in Multisig. I think we can also use this to unify the concept. "Get an extrinsic by its `extrinsicIndex` and block height or hash. The pair `blockId, extrinsicIndex` is sometimes referred to as a `Timepoint`."
https://github.com/paritytech/substrate/blob/d1923a4bc1eded37e57062f8b501c646e5321221/frame/multisig/src/lib.rs#L105

File: docs/src/openapi-v1.yaml [hunk 19]
```diff
@@ -273,6 +273,56 @@ paths:
             application/json:
               schema:
                 $ref: '#/components/schemas/Error'
+  /blocks/{blockId}/extrinsics/{extrinsicsIndex}:
```
Comment (joepetrowski) [inline docs/src/openapi-v1.yaml:276]:
```suggestion
  /blocks/{blockId}/extrinsics/{extrinsicIndex}:
```

File: src/services/blocks/BlocksService.spec.ts (hunk 16 above)
Comment (TarikGul) [reply src/services/blocks/BlocksService.spec.ts]: After a quick thought about this one, I will just change the index for this instead of adding an extra test that in essence will do the same exact thing. I like the insight though, and Ill make sure to comment that in the test itself.

File: src/controllers/blocks/BlocksExtrinsicsController.ts (hunk 13 above)
Comment (TarikGul) [reply src/controllers/blocks/BlocksExtrinsicsController.ts:49]: Yea I agree. `omitFinalizedTag` and `queryFinalizedHead`  are not connected to the output of this route, so we definitely don't need to assume the `this.options` might be fulfilled.

File: src/services/test-helpers/responses/blocks/block789629Extrinsic.json (hunk 18 above)
Comment (dvdplm) [reply src/services/test-helpers/responses/blocks/block789629Extrinsic.json]: I actually don't know why it's being signaled by github. Like why is not having a final LF a concern anyway? But there it is and my eye is drawn to "red thingies on screen", so might as well obey. :)

**Review by dvdplm — APPROVED** [review]

File: docs/src/openapi-v1.yaml (hunk 2 above)
Comment (TarikGul) [reply docs/src/openapi-v1.yaml]: @joepetrowski So would you think keeping the same name `getExtrinsicsByExtrsinsicIndex` for the new API route is the way to go, or changing it too `getExtrinsicsByTimepoint`?

File: src/controllers/blocks/BlocksExtrinsicsController.ts [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+			['/:extrinsicIndex', this.getExtrinsicByExtrsinsicIndex],
+		]);
+	}
+
+	/**
+	 *
+	 * @param _req Express Request
+	 * @param res Express Response
+	 */
+	private getExtrinsicByExtrsinsicIndex: RequestHandler<INumberParam> = async (
```
Comment (dvdplm) [inline src/controllers/blocks/BlocksExtrinsicsController.ts:25]:
```suggestion
	private getExtrinsicByIndex: RequestHandler<INumberParam> = async (
```
…seems less of a tongue twister?

File: docs/src/openapi-v1.yaml (hunk 2 above)
Comment (dvdplm) [reply docs/src/openapi-v1.yaml]: I'm warming up to `Timepoint` but if we go that route we should be prepared to shill it across the board.

Comment (TarikGul) [reply docs/src/openapi-v1.yaml]: Yup exactly. I totally agree, we would definitely have to make it completely consistent across the board.

Comment (joepetrowski) [reply docs/src/openapi-v1.yaml]: Yeah, I'm on board with `getExtrinsicByTimepoint`. (Note the _singular_ `Extrinsic`. You either get all _extrinsics_ from a block, but a `Timepoint` always refers to a single extrinsic.)

File: docs/src/openapi-v1.yaml [hunk 21]
```diff
@@ -273,6 +273,57 @@ paths:
...
               schema:
                 $ref: '#/components/schemas/Error'
+  /blocks/{blockId}/extrinsics/{extrinsicIndex}:
+    get:
+      tags:
+      - blocks
+      summary: Get an extrinsic by its extrinsicIndex and block height or hash. 
+      The pair blockId, extrinsicIndex is sometimes referred to as a Timepoint.
+      description: Returns a single extrinsic.
+      operationId: getExtrinsicByExtrsinsicIndex
```
Comment (emostov) [inline docs/src/openapi-v1.yaml:283]:
```suggestion
      operationId: getExtrinsicByIndex
```

**Review by emostov — APPROVED** [review]
LGTM modulo a nit

File: docs/src/openapi-v1.yaml (hunk 21 above)
Comment (TarikGul) [reply docs/src/openapi-v1.yaml:283]: Yea I like this as well. Will change throughout the code.

**Comment by gutenye** [comment]
@TarikGul Thank you very much!

**Comment by TarikGul** [comment]
@gutenye Of course, thank you for bringing up the issue.
